services/metadata/src/metadata/persistence/avu.clj

```diff
@@ -77,6 +114,35 @@
                            :modified_on (sqlfn now))))
     (where (select-keys avu [:id]))))
 
+(defn- update-valid-avu
+  "Updates an AVU, validating its given target_type and target_id match what's already in the database."
+  [user-id {:keys [id] :as avu}]
+  (when-let [existing-avu (when id (get-avu-by-id id))]
+    (when (not= (select-keys existing-avu [:target_type :target_id])
+                (select-keys avu          [:target_type :target_id]))
+      (throw+ {:type  :clojure-commons.exception/exists
+               :error "AVU already attached to another target item."
+               :avu   existing-avu}))
+    (update-avu user-id avu)
+    avu))
+
+(defn- find-matching-avu
+  "Finds an existing AVU by attribute, value, unit, target_type, target_id, and (if included) its ID."
+  [avu]
+  (let [required-keys [:attribute :value :unit :target_type :target_id]]
+    (when (every? (partial contains? avu) required-keys)
+      (first (select :avus
+                     (where (-> (select-keys avu (conj required-keys :id))
+                                (update :target_type db/->enum-val))))))))
+
+(defn add-or-update-avu
+  [user-id avu]
+  (if-let [matching-avu (find-matching-avu avu)]
+    matching-avu
+    (if-let [existing-avu (update-valid-avu user-id avu)]
+      existing-avu
+      (add-avus user-id [avu]))))
```

Comment:
This new `add-or-update-avu` function allows the new Set/Update AVU endpoints to add a new AVU with an ID that doesn't already exist in the database.

This function also allows the new Set/Update AVU endpoints to not mark AVUs as updated if they haven't changed (the `modified_on` or `modified_by` fields won't be updated), using the new `find-matching-avu` function.